rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,526 @@
+
+The contract interface would look like (more APIs could be added depending on the discussions on this RFC)
+
+```ts
+type ClusterMessagePayload = Serializable;
+
+interface BroadcastOptions {
+  /**
+   * If true, will also send the message to the worker that sent it.
+   * Defaults to false.
+   */
+  sendToSelf?: boolean;
+  /**
+   * If true, the message will also be sent to subscribers subscribing after the message was effectively sent.
+   * Defaults to false.
+   */
+  persist?: boolean;
+}
+
+export interface ClusteringServiceSetup {
+  /**
+   * Return true if clustering mode is enabled, false otherwise
+   */
+  isEnabled: () => boolean;
+  /**
+   * Return the current worker's id. In non-clustered mode, will return `1`
+   */
+  getWorkerId: () => number;
+  /**
+   * Broadcast a message to other workers.
+   * In non-clustered mode, this is a no-op
+   */
+  broadcast: (type: string, payload?: ClusterMessagePayload, options?: BroadcastOptions) => void;
+  /**
+   * Registered a handler for given `type` of IPC messages
+   * In non-clustered mode, this is a no-op that returns a no-op unsubscription callback.
+   */
+  addMessageHandler: (type: string, handler: MessageHandler) => MessageHandlerUnsubscribeFn;
+  /**
+   * Returns true if the current worker has been elected as the main one. In non-clustered mode, will return true
+   */
+  isMainWorker: () => boolean;
+}
+```
+
+## Cross-worker communication
+
+For some of our changes (such as the /status API, see below), we will need some kind of cross-worker communication. This 
+will need to pass through the coordinator, which will also serve as an 'event bus', or IPC forwarder.
+
+This IPC API will be exposed from the clustering service.
+
+```ts
+export interface ClusteringServiceSetup {
+  // [...]
+  broadcast: (type: string, payload?: ClusterMessagePayload, options?: BroadcastOptions) => void;
+  addMessageHandler: (type: string, handler: MessageHandler) => MessageHandlerUnsubscribeFn;
+}
+```
+
+Notes:
+-  to reduce clustered and non-clustered mode divergence, in non-clustered mode, these APIs would just be no-ops. 
+   It will avoid to force (most) code to check in which mode Kibana is running before calling them.
+-  We could eventually use an Observable pattern instead of a handler pattern to subscribe to messages.
+
+## Executing code on a single worker
+
+In some scenarios, we would like to have parts of the code executed only from a single process. 
+
+The SO migration would be a good example: 
+we don't need to have each worker try to perform the migration, and we'd prefer to have one performing/trying the migration, 
+and the others wait for it. Due to the architecture, we can't have the coordinator perform such single-process jobs, 
+as it doesn't actually run a Kibana server.
+
+There are various ways to address such use-cases. What seems to be the best compromise right now would be the concept of 
```

Comment:
Yea, the main worker will be a worker like the others, the only distinction will be this flag, that would allow plugins to execute arbitrary parts of the code only on a single worker if they need to.